packages/headless/src/controllers/facets/automatic-facet-builder/headless-automatic-facet-builder.ts

```diff
@@ -0,0 +1,95 @@
+import {configuration} from '../../../app/common-reducers';
+import {CoreEngine} from '../../../app/engine';
+import {SearchEngine} from '../../../app/search-engine/search-engine';
+import {setDesiredCount} from '../../../features/facets/automatic-facet-set/automatic-facet-set-actions';
+import {automaticFacetSetReducer as automaticFacetSet} from '../../../features/facets/automatic-facet-set/automatic-facet-set-slice';
+import {searchReducer as search} from '../../../features/search/search-slice';
+import {
+  AutomaticFacetSection,
+  ConfigurationSection,
+  SearchSection,
+} from '../../../state/state-sections';
+import {loadReducerError} from '../../../utils/errors';
+import {
+  Controller,
+  buildController,
+} from '../../controller/headless-controller';
+import {
+  AutomaticFacet,
+  buildAutomaticFacet,
+} from '../automatic-facet/headless-automatic-facet';
+
+export interface AutomaticFacetBuilder extends Controller {
+  /**
+   * The state of the `AutomaticFacetBuilder` controller.
+   */
+  state: AutomaticFacetBuilderState;
+}
+
+export interface AutomaticFacetBuilderState {
+  /**
+   * @beta - This property is part of the automatic facets feature.
+   * Automatic facets are currently in beta testing and should be available soon.
+   *
+   * The list of automatic facet controllers.
+   */
+  automaticFacets?: AutomaticFacet[];
+}
+
+export interface AutomaticFacetBuilderProps {
+  /**
+   * @beta - This property is part of the automatic facets feature.
+   * Automatic facets are currently in beta testing and should be available soon.
+   *
+   * The desired count of automatic facets.
+   * Must be a positive integer.
+   */
+  desiredCount: number;
+}
+
+/**
+ * @beta - This function is part of the automatic facets feature.
+ * Automatic facets are currently in beta testing and should be available soon.
+ *
+ * Creates a `AutomaticFacetBuilder` instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The automatic facets props. Automatic facets are currently in beta testing and should be available soon.
+ * @returns A `AutomaticFacetBuilder` controller instance.
+ */
+export function buildAutomaticFacetBuilder(
+  engine: SearchEngine,
+  props: AutomaticFacetBuilderProps
+): AutomaticFacetBuilder {
+  if (!loadAutomaticFacetBuilderReducers(engine)) {
+    throw loadReducerError;
+  }
+  const {dispatch} = engine;
+  dispatch(setDesiredCount(props.desiredCount));
+
+  const controller = buildController(engine);
+  const getAutomaticFacets = () =>
```

Comment:
Why is this a method that returns a value rather than just being the value itself? I would argue you could remove this logic and just update the `automaticFacets` assignment logic in the state getter to something like
```
const automaticFacets =
        engine.state.search.response.generateAutomaticFacets?.facets.map(
          (facet) => buildAutomaticFacet(engine, {field: facet.field})
        );
```

Or if you're not a fan of putting too much logic into a single statement, you could split it up as you're doing right now, just without making it a pointer to a method 🙂

Reply:
I agree it is kinda wierd. I got inspired by other controllers but this is clearer this way imo. I changed it.